File: UI/src/utils/unit-util.ts
import moment from 'moment';

export function carry(
  value: number,
  units: string[],
  hex: number,
  withUnit = true
) {
  const k = hex;
  const i = Math.floor(Math.log(value) / Math.log(k));
  if (withUnit) {
    return `${Math.round((value / k ** i) * 100) / 100} ${
      units[i < 0 ? 0 : i]
    }`;
  }
  return Math.round((value / k ** i) * 100) / 100;
}

/**
 * 处理uptime单位转换
 * @param seconds
 * @param withUnit
 */
export function uptimeFormatter(seconds: number, withUnit = true) {
  const days = Math.floor(seconds / 86400);
  seconds %= 86400;
  const hours = Math.floor(seconds / 3600);
  seconds %= 3600;
  const minutes = Math.floor(seconds / 60);
  seconds = Math.round(seconds % 60);

  return `${days}${withUnit ? '天' : ''}, ${hours
    .toString()
    .padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${seconds
    .toString()
    .padStart(2, '0')}`;
}

/**
 * 处理ms s单位转换
 * @param value
 * @param unit
 * @param withUnit
 */
export function formatTime(value: number, unit = 'ms', withUnit = true) {
  let totalSeconds;
  function formatNumber(num: number) {
    return num % 1 === 0 ? num.toString() : num.toFixed(2);
  }

  if (unit === 'ms') {
    totalSeconds = value / 1000;
  } else if (unit === 's') {
    totalSeconds = value;
    if (totalSeconds < 1) {
      const milliseconds = formatNumber(totalSeconds * 1000);
      return withUnit ? `${milliseconds}毫秒` : milliseconds;
    }
  } else {
    throw new Error('不支持的单位。只允许使用 "s" 和 "ms"。');
  }

  if (totalSeconds < 60) {
    if (totalSeconds < 0.1) {
      return withUnit ? `${value}毫秒` : value;
    }
    const result = formatNumber(totalSeconds);
    return withUnit ? `${result}秒` : result;
  }
  if (totalSeconds < 3600) {
    const minutes = Math.floor(totalSeconds / 60);
    const seconds = formatNumber(totalSeconds % 60);
    return withUnit
      ? `${minutes}分钟${seconds}秒`
      : `${minutes}分钟${seconds}秒`;
  }
  if (totalSeconds < 86400) {
    const hours = Math.floor(totalSeconds / 3600);
    totalSeconds %= 3600;
    const minutes = Math.floor(totalSeconds / 60);
    const seconds = formatNumber(totalSeconds % 60);
    return withUnit
      ? `${hours}小时${minutes}分钟${seconds}秒`
      : `${hours}小时${minutes}分钟${seconds}秒`;
  }
  const days = Math.floor(totalSeconds / 86400);
  totalSeconds %= 86400;
  const hours = Math.floor(totalSeconds / 3600);
  totalSeconds %= 3600;
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = formatNumber(totalSeconds % 60);
  return withUnit
    ? `${days}天${hours}小时${minutes}分钟${seconds}秒`
    : `${days}天${hours}小时${minutes}分钟${seconds}秒`;
}

const getTransUnit = (baseUnit: string) => {
  if (baseUnit === 's') {
    return '秒';
  }
  if (baseUnit === 'ms') {
    return '毫秒';
  }
  return '';
};

export function unitFormatter(
  baseUnit: string,
  value: number,
  withUnit = true
): string | number {
  let result = value || '-';
  // value必须为number类型,在此进行强制转换
  value = Number(value);
  if (value === 0) {
    // 感叹号在zabbix监控项单位中表示不转换
    if (baseUnit.startsWith('!')) {
      return `0${baseUnit.replace('!', ' ')}`;
    }

    return withUnit && !['text', 'uptime', 'unixtime'].includes(baseUnit)
      ? `0 ${getTransUnit(baseUnit)}`
      : '0';
  }
  if (baseUnit !== 'text') {
    result = (value || '-') + (baseUnit || '');
    // 感叹号在zabbix监控项单位中表示不转换
    if (baseUnit.startsWith('!')) {
      result = Math.round(value * 100) / 100 + baseUnit.replace('!', ' ');
    } else {
      switch (baseUnit) {
        case 'bps':
          result = carry(
            value,
            ['bps', 'Kbps', 'Mbps', 'Gbps', 'Tbps', 'Pbps'],
            1000,
            withUnit
          );
          break;
        case 'Bps':
          result = carry(
            value,
            ['Bps', 'KBps', 'MBps', 'GBps', 'TBps', 'PBps'],
            1000,
            withUnit
          );
          break;
        case 'HZ':
          result = carry(
            value,
            ['HZ', 'KHZ', 'MHZ', 'GHZ', 'THZ', 'PHZ'],
            1000,
            withUnit
          );
          break;
        case 'b':
          result = carry(
            value,
            ['b', 'Kb', 'Mb', 'Gb', 'Tb', 'Pb', 'Eb', 'Zb', 'Yb'],
            1000,
            withUnit
          );
          break;
        case 'B':
          result = carry(
            value,
            ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'],
            1024,
            withUnit
          );
          break;
        case 'b/s':
          result = carry(
            value,
            [
              'b/s',
              'Kb/s',
              'Mb/s',
              'Gb/s',
              'Tb/s',
              'Pb/s',
              'Eb/s',
              'Zb/s',
              'Yb/s',
            ],
            1000,
            withUnit
          );
          break;
        case 'B/s':
          result = carry(
            value,
            [
              'B/s',
              'KB/s',
              'MB/s',
              'GB/s',
              'TB/s',
              'PB/s',
              'EB/s',
              'ZB/s',
              'YB/s',
            ],
            1024,
            withUnit
          );
          break;
        case 'ms':
          result = formatTime(value, 'ms', withUnit);
          break;
        case 's':
          result = formatTime(value, 's', withUnit);
          break;
        case 'uptime':
          result = uptimeFormatter(value, withUnit);
          break;
        case 'unixtime':
          result = moment(value * 1000).format('YYYY-MM-DD HH:mm:ss');
          break;
        case '%':
          if (withUnit) {
            result = `${Math.round(Number(value) * 100) / 100} %`;
          } else {
            result = Math.round(Number(value) * 100) / 100;
          }
          break;
        case 'sectors':
          result = carry(
            value,
            ['sectors', 'Ksectors', 'Msectors', 'Gsectors', 'Tsectors'],
            1000,
            withUnit
          );
          break;
        case 'ips':
          result = carry(
            value,
            ['ips', 'Kips', 'Mips', 'Gips', 'Tips'],
            1000,
            withUnit
          );
          break;
        case '':
          result = Math.round(Number(value) * 100) / 100;
          break;
        default:
          result = withUnit
            ? `${Math.round(Number(value) * 100) / 100} ${baseUnit}`
            : value;
      }
    }
  }
  return result;
}

export function formatFixedNumber(num: number, decimals: number) {
  const formattedNumber = num.toFixed(decimals);
  const parts = formattedNumber.split('.');
  if (parts.length === 2) {
    const integerPart = parts[0];
    let decimalPart = parts[1];
    while (decimalPart.endsWith('0')) {
      decimalPart = decimalPart.slice(0, -1);
    }
    if (decimalPart === '') {
      return integerPart;
    }
    return `${integerPart}.${decimalPart}`;
  }
  return formattedNumber;
}

File: crm-core-apps/src/main/java/cn/oli/crm/utils/UnitUtil.java
package cn.oli.crm.utils;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.DecimalFormat;
import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;

/**
 * 单位转换工具类
 */
public class UnitUtil {

    public static String carry(double value, String[] units, int hex, boolean withUnit) {
        int i = value == 0 ? 0 : (int) Math.floor(Math.log(value) / Math.log(hex));
        double scaledValue = round(value / Math.pow(hex, i));
        if (withUnit) {
            return scaledValue + " " + units[Math.max(i, 0)];
        }
        return String.valueOf(scaledValue);
    }

    private static double round(double value) {
        BigDecimal bd = BigDecimal.valueOf(value);
        bd = bd.setScale(2, RoundingMode.HALF_UP);
        return bd.doubleValue();
    }

    public static String formatUnixTime(long unixTime) {
        Instant instant = Instant.ofEpochSecond(unixTime);
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
                .withZone(ZoneId.systemDefault());
        return formatter.format(instant);
    }

    public static String uptimeFormatter(int seconds, boolean withUnit) {
        int days = seconds / 86400;
        seconds %= 86400;
        int hours = seconds / 3600;
        seconds %= 3600;
        int minutes = seconds / 60;
        seconds %= 60;

        if (withUnit) {
            return days + "天, " + String.format("%02d:%02d:%02d", hours, minutes, seconds);
        } else {
            return String.format("%d, %02d:%02d:%02d", days, hours, minutes, seconds);
        }
    }

    private static String formatNumber(double number, String unit) {
        DecimalFormat df = new DecimalFormat("#.##");
        return df.format(number) + unit;
    }


    public static String formatTime(double value, String unit, boolean withUnit) {
        double totalSeconds;
        if ("ms".equals(unit)) {
            totalSeconds = value / 1000.0;
        } else if ("s".equals(unit)) {
            totalSeconds = value;
            if (totalSeconds < 1) {
                return formatNumber(totalSeconds * 1000, withUnit ? "毫秒" : "");
            }
        } else {
            throw new IllegalArgumentException("Unsupported unit. Only \"s\" and \"ms\" are allowed.");
        }

        if (totalSeconds < 60) {
            return formatNumber(totalSeconds, withUnit ? "秒" : "");
        } else if (totalSeconds < 3600) {
            int minutes = (int) (totalSeconds / 60);
            double seconds = totalSeconds % 60;
            return withUnit
                    ? String.format("%d分钟%s秒", minutes, formatNumber(seconds, ""))
                    : String.format("%d:%s", minutes, formatNumber(seconds, ""));
        } else if (totalSeconds < 86400) {
            int hours = (int) (totalSeconds / 3600);
            int minutes = (int) ((totalSeconds % 3600) / 60);
            double seconds = totalSeconds % 60;
            return withUnit
                    ? String.format("%d小时%d分钟%s秒", hours, minutes, formatNumber(seconds, ""))
                    : String.format("%d:%02d:%s", hours, minutes, formatNumber(seconds, ""));
        } else {
            int days = (int) (totalSeconds / 86400);
            int hours = (int) ((totalSeconds % 86400) / 3600);
            return withUnit
                    ? String.format("%d天%d小时", days, hours)
                    : String.format("%d:%02d", days, hours);
        }
    }

    public static String unitFormatter(String baseUnit, double value, boolean withUnit) {
        switch (baseUnit) {
            case "bps":
                return carry(value, new String[]{"bps", "Kbps", "Mbps", "Gbps", "Tbps", "Pbps"}, 1000, withUnit);
            case "Bps":
                return carry(value, new String[]{"Bps", "KBps", "MBps", "GBps", "TBps", "PBps"}, 1000, withUnit);
            case "HZ":
                return carry(value, new String[]{"HZ", "KHZ", "MHZ", "GHZ", "THZ", "PHZ"}, 1000, withUnit);
            case "b":
                return carry(value, new String[]{"b", "Kb", "Mb", "Gb", "Tb", "Pb", "Eb", "Zb", "Yb"}, 1000, withUnit);
            case "B":
                return carry(value, new String[]{"B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"}, 1024, withUnit);
            case "b/s":
                return carry(value, new String[]{"b/s", "Kb/s", "Mb/s", "Gb/s", "Tb/s", "Pb/s", "Eb/s", "Zb/s", "Yb/s"}, 1000, withUnit);
            case "B/s":
                return carry(value, new String[]{"B/s", "KB/s", "MB/s", "GB/s", "TB/s", "PB/s", "EB/s", "ZB/s", "YB/s"}, 1024, withUnit);
            case "ms":
                return formatTime(value, "ms", withUnit);
            case "s":
                return formatTime(value, "s", withUnit);
            case "uptime":
                return uptimeFormatter((int) value, withUnit);
            case "unixtime":
                return formatUnixTime((long) value);
            case "%":
                return withUnit ? round(value) + "%" : String.valueOf(round(value));
            default:
                return withUnit ? round(value) + baseUnit : String.valueOf(round(value));
            // throw new IllegalArgumentException("Unsupported base unit: " + baseUnit);
        }
    }
}
